File: flutter/test_app/lib/data/firebase/repository/firebase_training_repository_impl.dart
import '../../../domain/model/common_models.dart';
import '../../../domain/model/prijavljen_trening.dart';
import '../../../domain/model/training_models.dart';
import '../../../domain/model/trener_models.dart';
import '../../../domain/repository/training_repository.dart';

import '../model/dtos.dart';
import '../model/firebase_mapper.dart';
import '../network/service/training_service.dart';

class FirebaseTrainingRepositoryImpl implements TrainingRepository {
  final FirebaseTrainingService service;

  FirebaseTrainingRepositoryImpl(this.service);

  @override
  Future<List<PrijavljenTrening>> getTrainingsForUser(String sportasId) async {
    try {
      final items = await service.myTrainingsRaw();
      return items.map(FirebaseMapper.myTrainingsItemToDomain).toList();
    } catch (error) {
      return [];
    }
  }

  @override
  Future<List<Teretana>> getTeretane() async {
    final dtoList = await service.getTeretane();
    return dtoList.map(FirebaseMapper.teretanaToDomain).toList();
  }

  @override
  Future<List<DvoranaSimple>> getDvorane() async {
    final dtoList = await service.getDvorane();
    return dtoList.map(FirebaseMapper.dvoranaToDomain).toList();
  }

  @override
  Future<List<VrstaTreninga>> getVrsteTreninga() async {
    final dtoList = await service.getVrsteTreninga();
    return dtoList.map(FirebaseMapper.vrstaToDomain).toList();
  }

  @override
  Future<List<VrstaTreningaOption>> getVrsteTreningaOptions() async {
    final dtoList = await service.getVrsteTreninga();
    return dtoList.map(FirebaseMapper.vrstaToOptionDomain).toList();
  }

  @override
  Future<List<TreningOption>> getTreningOptions() async {
    final dtos = await service.getTreningOptions();
    return dtos.map(FirebaseMapper.treningOptionToDomain).toList();
  }

  @override
  Future<List<DostupniTrening>> getTrainingsByDateAndTeretana({
    required String teretanaId,
    required DateTime date,
  }) async {
    final req = TrainingsByGymDateRequestDto(
      teretanaId: teretanaId,
      date: _dateToYyyyMmDd(date),
    );

    final dto = await service.trainingsByGymDate(req);
    return dto.items.map(FirebaseMapper.availableTrainingToDomain).toList();
  }

  @override
  Future<SignUpResult> signUpForTraining({
    required String rasporedId,
    required String korisnikId,
  }) async {
    try {
      final req = SignupForTrainingRequestDto(rasporedId: rasporedId);
      final resp = await service.signupForTraining(req);
      return FirebaseMapper.signupResultToDomain(resp);
    } catch (_) {
      return SignUpResult.error;
    }
  }

  @override
  Future<List<TrainerTraining>> getTrainingsForTrainer(String trenerId) async {
    final dto = await service.trainerTrainings();
    return dto.items.map(FirebaseMapper.trainerTrainingToDomain).toList();
  }

  @override
  Future<List<PrijavljeniSudionik>> getAttendeesForRaspored(String rasporedId) async {
    final req = GetAttendeesRequestDto(rasporedId: rasporedId);
    final dto = await service.attendeesByRaspored(req);
    return dto.items.map(FirebaseMapper.attendeeToDomain).toList();
  }

  @override
  Future<AttendanceUpdateResult> setAttendanceForRaspored({
    required String rasporedId,
    required List<AttendanceUpdate> updates,
  }) async {
    final req = FirebaseMapper.toAttendanceRequestDto(
      rasporedId: rasporedId,
      updates: updates,
    );

    final dto = await service.setAttendanceForRaspored(req);
    return FirebaseMapper.attendanceResultToDomain(dto);
  }

  @override
  Future<void> createTrening(CreateTreningInput input) async {
    final req = FirebaseMapper.toCreateTrainingRequestDto(input);
    await service.createTraining(req);
  }

  @override
  Future<void> createRaspored(CreateRasporedInput input) async {
    final req = FirebaseMapper.toCreateRasporedRequestDto(input);
    await service.createRaspored(req);
  }

  @override
  Future<DeleteRasporedResult> deleteRaspored(String rasporedId) async {
    final req = DeleteRasporedRequestDto(rasporedId: rasporedId);
    final dto = await service.deleteRaspored(req);
    return FirebaseMapper.deleteRasporedToDomain(dto);
  }

  String _dateToYyyyMmDd(DateTime d) {
    final u = DateTime(d.year, d.month, d.day);
    final mm = u.month.toString().padLeft(2, '0');
    final dd = u.day.toString().padLeft(2, '0');
    return '${u.year}-$mm-$dd';
  }
}

File: kotlin_jetpack_compose/app/src/main/java/com/example/testapp/data/firebase/repository/FirebaseTrainingRepositoryImpl.kt
package com.example.testapp.data.firebase.repository

import com.example.testapp.data.firebase.model.mapper.toDomain
import com.example.testapp.data.firebase.model.mapper.toDto
import com.example.testapp.data.firebase.network.service.FirebaseTrainingService
import com.example.testapp.data.supabase.model.DTOs.toDomain
import com.example.testapp.domain.model.AttendanceUpdate
import com.example.testapp.domain.model.AttendanceUpdateResult
import com.example.testapp.domain.model.CreateRasporedRequest
import com.example.testapp.domain.model.CreateTreningRequest
import com.example.testapp.domain.model.CreateTreningResult
import com.example.testapp.domain.model.DeleteResult
import com.example.testapp.domain.model.DostupniTrening
import com.example.testapp.domain.model.Dvorana
import com.example.testapp.domain.model.PrijavljenTrening
import com.example.testapp.domain.model.PrijavljeniSudionik
import com.example.testapp.domain.model.Raspored
import com.example.testapp.domain.model.Teretana
import com.example.testapp.domain.model.TrainingDetails
import com.example.testapp.domain.model.TrenerTrening
import com.example.testapp.domain.model.TreningOption
import com.example.testapp.domain.model.VrstaTreninga
import com.example.testapp.domain.repository.TrainingRepository
import com.google.firebase.auth.FirebaseAuth
import java.time.LocalDate
import java.time.LocalDateTime
import javax.inject.Inject
import kotlin.collections.map

class FirebaseTrainingRepositoryImpl @Inject constructor(
    private val trainingService: FirebaseTrainingService,
    private val firebaseAuth: FirebaseAuth
) : TrainingRepository {

    override suspend fun getTrainingsForUser(sportasId: String): List<PrijavljenTrening> {



        return trainingService.getTrainingsForUser(sportasId).map { it.toDomain() }
    }

    override suspend fun getTrainingsByDateAndTeretana(teretana: String, date: LocalDate): List<DostupniTrening> {
        return trainingService.getTrainingsByDateAndTeretana(teretanaId = teretana, date = date)
    }

    override suspend fun prijaviSeNaTrening(rasporedId: String) {
        val uid = firebaseAuth.currentUser?.uid
            ?: throw IllegalStateException("Korisnik nije prijavljen.")
        trainingService.signupForTraining(korisnikId = uid, rasporedId = rasporedId)
    }

    override suspend fun getTrainingDetails(treningId: String): TrainingDetails {
        return trainingService.getTrainingDetails(treningId)
    }

    override suspend fun getDvorane(): List<Dvorana> {
        return trainingService.getDvorane().map { it.toDomain() }
    }

    override suspend fun getTeretane(): List<Teretana> {
        return trainingService.getTeretane().map { it.toDomain() }
    }

    override suspend fun getTrainingsForTrainer(trenerId: String): List<TrenerTrening> {
        return trainingService.getTrainingsForTrainer(trenerId)
    }

    override suspend fun getAttendeesForRaspored(rasporedId: String): List<PrijavljeniSudionik> {
        return trainingService.getAttendeesByRaspored(rasporedId)
    }

    override suspend fun setAttendanceForRaspored(
        rasporedId: String,
        updates: List<AttendanceUpdate>
    ): AttendanceUpdateResult {

        return trainingService.setAttendanceForRaspored(rasporedId, updates)
    }

    override suspend fun getVrsteTreninga(): List<VrstaTreninga> {
        return trainingService.getVrsteTreninga().map { it.toDomain() }
    }

    override suspend fun createTrening(request: CreateTreningRequest): CreateTreningResult {

        return trainingService.createTrening(request)
    }

    override suspend fun getTreningOptions(): List<TreningOption> {
        return trainingService.getTreningOptionsDto().toDomain()
    }

    override suspend fun createRaspored(request: CreateRasporedRequest): Raspored {
        val dto = com.example.testapp.data.firebase.model.dto.CreateRasporedRequestDto(
            treningId = request.treningId,
            trenerId = request.trenerId,
            pocetak = request.pocetak.toString(),
            zavrsetak = request.zavrsetak.toString(),
            idRasporeda = request.idRasporeda
        )

        val created = trainingService.createRaspored(request)
        return Raspored(
            idRasporeda = created.idRasporeda,
            treningId = created.treningId,
            trenerId = created.trenerId,
            pocetak = created.pocetak,
            zavrsetak = created.zavrsetak
        )
    }

    override suspend fun deleteRaspored(rasporedId: String): DeleteResult {
        return trainingService.deleteRaspored(rasporedId)
    }
}
